src/theory/strings/theory_strings.cpp

```diff
@@ -4110,51 +4088,57 @@ Node TheoryStrings::mkLength( Node t ) {
   return Rewriter::rewrite( NodeManager::currentNM()->mkNode( kind::STRING_LENGTH, t ) );
 }
 
-Node TheoryStrings::mkExplain( std::vector< Node >& a ) {
+Node TheoryStrings::mkExplain(const std::vector<Node>& a)
+{
   std::vector< Node > an;
   return mkExplain( a, an );
 }
 
-Node TheoryStrings::mkExplain( std::vector< Node >& a, std::vector< Node >& an ) {
+Node TheoryStrings::mkExplain(const std::vector<Node>& a,
+                              const std::vector<Node>& an)
+{
   std::vector< TNode > antec_exp;
-  for( unsigned i=0; i<a.size(); i++ ) {
-    if( std::find( a.begin(), a.begin() + i, a[i] )==a.begin() + i ) {
-      bool exp = true;
-      Debug("strings-explain") << "Ask for explanation of " << a[i] << std::endl;
-      //assert
-      if(a[i].getKind() == kind::EQUAL) {
-        //Assert( hasTerm(a[i][0]) );
-        //Assert( hasTerm(a[i][1]) );
-        Assert( areEqual(a[i][0], a[i][1]) );
-        if( a[i][0]==a[i][1] ){
-          exp = false;
-        }
-      } else if( a[i].getKind()==kind::NOT && a[i][0].getKind()==kind::EQUAL ) {
-        Assert( hasTerm(a[i][0][0]) );
-        Assert( hasTerm(a[i][0][1]) );
-        AlwaysAssert( d_equalityEngine.areDisequal(a[i][0][0], a[i][0][1], true) );
-      }else if( a[i].getKind() == kind::AND ){
-        for( unsigned j=0; j<a[i].getNumChildren(); j++ ){
-          a.push_back( a[i][j] );
-        }
-        exp = false;
-      }
-      if( exp ) {
-        unsigned ps = antec_exp.size();
-        explain(a[i], antec_exp);
-        Debug("strings-explain") << "Done, explanation was : " << std::endl;
-        for( unsigned j=ps; j<antec_exp.size(); j++ ) {
-          Debug("strings-explain") << "  " << antec_exp[j] << std::endl;
-        }
-        Debug("strings-explain") << std::endl;
+  // copy to processing vector
+  std::vector<Node> ap;
+  ap.insert(ap.end(), a.begin(), a.end());
+  for (unsigned i = 0; i < ap.size(); i++)
+  {
+    Node api = ap[i];
+    if (std::find(ap.begin(), ap.begin() + i, api) != ap.begin() + i)
+    {
+      // already processed
+      continue;
+    }
+    if (api.getKind() == AND)
+    {
+      for (const Node& apic : api)
+      {
+        ap.push_back(apic);
```

Comment:
I am not a huge fan of modifying `ap` while iterating over it. The for loop above makes it seem like the size is fixed. Why not do two loops? Instead of copying `a` to `ap` in the beginning, iterate over `ap` and flatten the `AND`s into `ap`. This shouldn't be less efficient (in fact you save copying some useless elements into `ap`).

Reply:
The flattening of AND may need to be done recursively, which makes this a bit of a pain.  I'll make a utility to do this.

Reply:
Yeah, but it is worth it IMHO :) The updated code is much easier to read.